### PR TITLE
CLC-6250 Centralize delegate logic

### PR DESCRIPTION
**Pull request by raydavis**
https://jira.ets.berkeley.edu/jira/browse/CLC-6250

### CODE REVIEW

File: spec/models/my_academics/filtered_for_delegate_spec.rb [hunk 1]
```diff
@@ -65,6 +53,7 @@
     include_examples 'filtered feed'
 
     it 'should return grades' do
+      pp feed
```
Comment (paulkerschen) [inline spec/models/my_academics/filtered_for_delegate_spec.rb]: @raydavis Found one!

Comment (raydavis) [inline spec/models/my_academics/filtered_for_delegate_spec.rb]: My old nemesis!

File: src/assets/javascripts/angular/controllers/widgets/statusController.js [hunk 2]
```diff
@@ -111,12 +111,19 @@ angular.module('calcentral.controllers').controller('StatusController', function
 
       // Get all the necessary data from the different factories
       var getBadges = badgesFactory.getBadges().success(loadStudentInfo);
-      var getFinances = financesFactory.getFinances().success(loadFinances);
-      var getFinaidActivityOld = activityFactory.getFinaidActivityOld().then(loadActivity);
       var getHolds = holdsFactory.getHolds().then(loadHolds);
+      var statusGets = [getBadges, getHolds];
```
Comment (raydavis) [inline src/assets/javascripts/angular/controllers/widgets/statusController.js]: @christianvuerings , this works in my tests but I always appreciate your reviews.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -48,6 +49,16 @@ def check_reauthentication
...
   end
 
+  # Only a small subset of student API feeds are available to a delegate, and so
+  # these methods filter controller endpoints by default.
+  def accessible_by_delegate?
+    false
+  end
+  def deny_if_filtered
+    filtered = !accessible_by_delegate? && current_user.authenticated_as_delegate?
+    raise Pundit::NotAuthorizedError.new("By delegate #{current_user.original_delegate_user_id}") if filtered
```
Comment (johncrossman) [inline app/controllers/application_controller.rb]: @raydavis Any value in knowing which student UID the delegate is viewing-as when NotAuthorized is raised?

File: src/assets/javascripts/angular/controllers/widgets/statusController.js (hunk 2 above)
Comment (christianvuerings) [inline src/assets/javascripts/angular/controllers/widgets/statusController.js]: @raydavis this lgtm

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (raydavis) [inline app/controllers/application_controller.rb]: Yes, but that turns out to be provided by the built-in error handling. E.g.,
`[2016-03-21 16:53:36] [WARN] Unauthorized request made by UID: 76447 to my_groups#get_feed: By delegate 43667`

**Comment by johncrossman** [comment]
Looks good, @raydavis  Thank you VERY much for getting this done so quickly... and providing more elegance to the authorization logic.

**Comment by paulkerschen** [comment]
Looks good to me too.

**Pull request merged**